docs/cpp/source/notes/tensor_cuda_stream.rst

```diff
@@ -0,0 +1,173 @@
+
+.. code-block:: cpp
+
+  void setCurrentCUDAStream(CUDAStream stream);
+
+.. attention::
+
+  This function may have nosthing to do with the current device. It only changes the current stream on the stream's device.
+  We recommend using ``CUDAStreamGuard``, instead, since it switches to the stream's device and makes it the current stream on that device.
+  ``CUDAStreamGuard`` will also restore the current device and stream when it's destroyed
+
+2. Use ``CUDAStreamGuard`` to switch to a CUDA stream within a scope, it is defined in `CUDAStreamGuard.h`_
+
+.. tip::
+
+  Use ``CUDAMultiStreamGuard`` if you need to set streams on multiple CUDA devices.
+
+CUDA Stream Usage Examples
+**************************
+
+1. Acquiring and setting a CUDA stream
+
+.. code-block:: cpp
+
+  // get a new stream from CUDA stream pool on current device (device 0)
+  // set current stream with this new stream and get current stream
+  torch::cuda::CUDAStream myStream = at::cuda::getStreamFromPool();
+  torch::cuda::setCurrentCUDAStream(myStream);
+  torch::cuda::CUDAStream curStream = torch::cuda::getCurrentCUDAStream();
+  assert(curStream == myStream);
+
+  // get the default CUDA stream on current device (device 0)
+  // set current stream with default stream and get current stream
+  torch::cuda::CUDAStream defaultStream = torch::cuda::getDefaultCUDAStream();
+  torch::cuda::setCurrentCUDAStream(defaultStream);
+  curStream = torch::cuda::getCurrentCUDAStream();
+  assert(curStream == defaultStream);
+  assert(myStream != defaultStream)
+
+.. attention::
+  
+  Above code is running on the same device/gpu (say device 0). `setCurrentCUDAStream` works as expected and set current CUDA stream correctly
+  on device 0, this is because all the streams are on the same device. However, say above code is running on device 0 and we get `myStream`
+  from device 1, then `torch::cuda::setCurrentCUDAStream(myStream)` will set `myStream` as current stream on device 1, not device 0.
+
+
+2. Use various CUDA Guard to set CUDA stream
+
+.. code-block:: cpp
+
+   // create a vector of CUDA stream from current device (device 0)
+   std::vector<torch::cuda::CUDAStream> streams0 = {torch::cuda::getDefaultCUDAStream(), torch::cuda::getStreamFromPool()};
+   assert(streams0[0].device_index() == 0);
+   assert(streams0[1].device_index() == 0);
+   // set current stream as `streams0[0]` on `streams0[0]`'s current device (device 0)
+   torch::cuda::setCurrentCUDAStream(streams0[0]);
+
+   // create a vector of CUDA stream from device 1
+   std::vector<torch::cuda::CUDAStream> streams1;
+   {
+     // create a CUDA Device guard within the scope to guard device 1
+     torch::cuda::CUDAGuard device_guard(1);
+     // get default CUDA stream from device 1
+     streams1.push_back(torch::cuda::getDefaultCUDAStream());
+     // get a new stream from CUDA stream pool on device 1
+     streams1.push_back(torch::cuda::getStreamFromPool());
+   }
```

Comment:
You may want to add a comment explaining what the device guard is doing here. That is, it switches to device 1, then when it goes out of scope it resets the current device to device 0.

Reply:
I think previous comments are enough, but I will add these.

> This is a helpful document for people trying to use streams with the C++ API. The one part of it that may need a change is the examples. Maybe something like:
> 
> * a combination of your first two examples, where you acquire a stream, switch to it with setCurrentCUDAStream, say that you'd do some work here, and then switch back to the default stream. Also, do you want to show events here, too?
> * an example where you use a stream guard to do the same thing
> * an example where you use a stream guard. to also change the device
> * an example where you use a multistream guard
> 
> One narrative to motivate the multi-device setting may just be "now suppose you want to do work on the non-default stream on two devices simultaneously, and we already have streams on both devices in a vector, then..."
> 
> What are your thoughts?

Yes, I can do it. 
I think there is a misunderstanding here :).  I was not aiming to make this as a tutorial, but a centralized tech doc for programming people to use CUDA stream. So i focused on what action will result in what behavior, user should be able to mixed these actions to meet their needs.

I get your point, I'll re-org the examples and add more examples as you suggested, I think that make sense.